carddav/carddav.go

```diff
@@ -263,7 +263,143 @@ func (b *backend) ListAddressObjects(req *carddav.AddressDataRequest) ([]carddav
 }
 
 func (b *backend) QueryAddressObjects(query *carddav.AddressBookQuery) ([]carddav.AddressObject, error) {
-	panic("TODO")
+	all, err := b.ListAddressObjects(&query.DataRequest)
+	if err != nil {
+		return nil, fmt.Errorf("could not list address objects: %w", err)
+	}
+
+	if query == nil {
+		// no query, return the full list of contacts.
+		return all, nil
+	}
+
+	var aos []carddav.AddressObject
+	if query.Limit > 0 {
+		aos = make([]carddav.AddressObject, 0, query.Limit)
+	}
+
+loop:
+	for _, ao := range all {
+		ok, err := match(query, ao)
+		if err != nil {
+			return nil, err
+		}
+		if ok {
+			aos = append(aos, ao)
+			if query.Limit > 0 && len(aos) >= query.Limit {
+				break loop
+			}
+		}
+	}
+
+	return aos, nil
+}
+
+func match(query *carddav.AddressBookQuery, ao carddav.AddressObject) (bool, error) {
+	if query.DataRequest.AllProp {
+		for _, name := range query.DataRequest.Props {
+			field := ao.Card.Get(name)
+			if field == nil {
+				// missing required property.
+				return false, fmt.Errorf("missing property %q", name)
+			}
+		}
+	}
+
+	switch query.FilterTest {
+	default:
+		return false, fmt.Errorf("unknown query filter test %q", query.FilterTest)
+
+	case carddav.FilterAnyOf, "":
+		for _, prop := range query.PropFilters {
+			ok, err := matchPropFilter(prop, ao)
+			if err != nil {
+				return false, err
+			}
+			if ok {
+				return true, nil
+			}
+		}
+		return false, nil
+
+	case carddav.FilterAllOf:
+		for _, prop := range query.PropFilters {
+			ok, err := matchPropFilter(prop, ao)
+			if err != nil {
+				return false, err
+			}
+			if !ok {
+				return false, nil
+			}
+		}
+		return true, nil
+	}
+}
+
+func matchPropFilter(prop carddav.PropFilter, ao carddav.AddressObject) (bool, error) {
+	field := ao.Card.Get(prop.Name)
+	if field == nil {
+		// assume AddressBookQuery.DataRequest.AllProp is false
```

Comment:
`IsNotDefined` probably means that we need to check `field == nil`. Would need to read again the RFC to double check.